Linux/sgx/test_app/enclave/tests/evp_smx.c

```diff
@@ -36,591 +36,666 @@
-            ret = -4;
-            break;			
-        }
-        if (!PEM_write_bio_ECPrivateKey(pri_bio, ec_key, NULL, NULL, 0, NULL, NULL)) {
-            printf("Error: fail to write private key from ec_key to the BIO\n");
-            ret = -5;
-            break;			
-        }
-        pri_len = BIO_pending(pri_bio);
-        if (pri_len == 0) {
-            printf("Error: fail to get size of the BIO for private key\n");
-            ret = -6;
-            break;
-        }
-        *private_key = (char*)malloc(pri_len+1);
-        if (BIO_read(pri_bio, *private_key, pri_len) <= 0) {
-            printf("Error: fail to read private key from the BIO\n");
-            ret = -7;
-            break;
-        }
-        (*private_key)[pri_len] = '\0';
-
-        // 5. Generate SM2 public key based on the curve
-        pub_bio = BIO_new(BIO_s_mem());
-        if (pub_bio == NULL) {
-            printf("Error: fail to create a BIO for public key\n");
-            ret = -8;
-            break;
-        }
-        if (!PEM_write_bio_EC_PUBKEY(pub_bio, ec_key)) {
-            printf("Error: fail to write public key from ec_key to the BIO\n");
-            ret = -9;
-            break;
-        }
-        pub_len = BIO_pending(pub_bio);
-        if (pub_len == 0) {
-            printf("Error: fail to get size of the BIO for public key\n");
-            ret = -10;
-            break;
-        }
-        *public_key = (char*)malloc(pub_len+1);
-        if (BIO_read(pub_bio, *public_key, pub_len) <= 0) {
-            printf("Error: fail to read public key from the BIO\n");
-            ret = -11;
-            break;
-        }
-        (*public_key)[pub_len] = '\0';
-
-    } while(0);
-
-    // 6. Finalize
-    if (ret != 0)
-    {
-        SAFE_FREE(*private_key, pri_len+1);
-        SAFE_FREE(*public_key, pub_len+1);
-    }	
-	EC_KEY_free(ec_key);
+	int ret = 0;
+	BIO *pri_bio = NULL, *pub_bio = NULL;
+	size_t pri_len = 0, pub_len = 0;
+	EVP_PKEY *evp_pkey = NULL;
+	EVP_PKEY_CTX *evp_pkey_ctx = NULL;
+
+	do {
+#if OPENSSL_VERSION_NUMBER < 0x30000000
+		printf("OpenSSL version: 1.1.1 or lower\n");
+		// 1. Create an EC_GROUP object with a curve specified by SM2 NID
+		EC_GROUP *ec_group = NULL;
+		EC_KEY *ec_key = NULL;
+		ec_group = EC_GROUP_new_by_curve_name(NID_sm2);
+		if (ec_group == NULL) {
+			printf("Error: fail to create an EC_GROUP object for SM2\n");
+			ret = -1;
+			break;
+		}
+		// 2. Create a new EC key
+		ec_key = EC_KEY_new();
+		if (ec_key == NULL) {
+			printf("Error: fail to create a new EC key\n");
+			ret = -2;
+			break;
+		}
+		// 3. Set the new EC key's curve
+		if (EC_KEY_set_group(ec_key, ec_group) != 1) {
+			printf("Error: fail to set the new EC key's curve\n");
+			ret = -3;
+			break;
+		}
+		// 4. Generate key pair based on the curve
+		if (!EC_KEY_generate_key(ec_key)) {
+			printf("Error: fail to generate key pair based on the curve\n");
+			ret = -4;
+			break;
+		}
+#elif OPENSSL_VERSION_NUMBER >= 0x30000000
```

Comment:
You can simply use only `#else` here, since it should be enough when it is `#if OPENSSL_VERSION_NUMBER < 0x30000000` above.